### DISCUSSION TITLE
Clickhouse

### DISCUSSION THREAD

**Discussion by flaneur2020** [evidence-dev/evidence#2339] [Data Source Request]
**Feature Description**

Clickhouse is a widely used analystic databases, evidence can help making report with the data which stored in clickhouse or any clickhouse-compatible alternatives.

**Goal of Feature**
- making report with data in clickhouse or any clickhouse-compatible alternative.

**Current Solution**
- nope

**Are you interested in contributing this feature?**
- yes, but i'm not fimiliar with the frontend tool chain. If there's a guide about how to contribute another data source, i'd like to contribute this feature.

**Comment by archiewood** [comment]
I think Clickhouse is already supported via the Postgres connector. See #181.

Can you test to confirm? (We don't have a Clickhouse db!)

**Comment by flaneur2020** [comment]
> I think Clickhouse is already supported via the Postgres connector. See https://github.com/evidence-dev/evidence/issues/181.

IMO, clickhouse have a postgres interface compatibility layer, but in practise it's less popular than its Native protocol and HTTP protocol. 

the PR #516 takes advantage of the npm package https://github.com/TimonKK/clickhouse to leverage the support about clickhouse with the 8123 port. This port is exposed in the instances created in Clickhouse Cloud service (however, it did not exposes the postgres protocol's port).

> Can you test to confirm? (We don't have a Clickhouse db!)

sure! can you give a hand about how to add the test on clickhouse into the ci? or can i test with the sites/test-env/pages/firstquery.md?

**Comment by flaneur2020** [comment]
Here is the output with the clickhouse data connection in `sites/test-env`:

![Screen Shot 2022-12-13 at 22 08 30](https://user-images.githubusercontent.com/129800/207355273-308a7768-605f-4680-bd30-3396cd99d69e.png)

**Comment by archiewood** [comment]
> IMO, clickhouse have a postgres interface compatibility layer, but in practise it's less popular than its Native protocol and HTTP protocol.

Gotcha, makes sense.

> sure! can you give a hand about how to add the test on clickhouse into the ci? or can i test with the sites/test-env/pages/firstquery.md?

This needs a bit more thought. 

The way we run the CI with existing DBs is against our own DB of that type. We configure the connection using environment variables in the github.

To date we've covered the DBs we've got most experience using in analytics, and that our cloud customers are using.

## Rethinking connectors

As we go forward we want the community to be able to use their other databases, but it's impractical / uneconomical for us to spin up every database type for testing. There's also a question of maintainability for the core team.

On this basis, we're looking to re-architect evidence so that community members can plug in their own connectors that they maintain. See https://github.com/evidence-dev/evidence/issues/486

I think Clickhouse probably falls in this category for us: want to support it, but not spend core team time maintaining it.

**Would you be up for starting Clickhouse as one of the first "community supported" connectors?**

What this would involve in practise:
- Wait for us to rearchitect to support this - hoping to get this done early 2023.
- Publishing your own package on npm - probably from a separate repo you maintain. (We'd be able to provide a template for this at that time)

**Comment by flaneur2020** [comment]
sure, looking forward for the new connectors architecture!

**Comment by archiewood** [comment]
Hello - looping back on connector requests

We have recently re-architected Evidence (since this issue was opened), and Evidence is now "pluggable", meaning you can write your own connectors, publish them, and use them in Evidence.

If you are interesting in writing a connector, here are some resources:

1. Docs on creating a plugin: https://docs.evidence.dev/plugins/creating-a-plugin/datasources
2. Template for a connector plugin: https://github.com/evidence-dev/datasource-template
3. An example that I spun up for pulling stats from the npm API: https://github.com/archiewood/npm-stats

If you do write a plugin:

- the suggested package name is `evidence-connector-[plugin]`
- the docs are new, so let us know if there is extra information you need
- share it on the Evidence [community](https://github.com/evidence-dev/evidence/issues/slack.evidence.dev), and we'll feature it!
